File: tray.py
#!/usr/bin/env python3
"""System tray icon for sonos-linux. Reads JSON from stdin, prints commands to stdout."""

import json
import signal
import sys

import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk, GLib


class SonosTray:
    def __init__(self, data):
        self.data = data

        self.icon = Gtk.StatusIcon()
        self.icon.set_from_icon_name("audio-speakers")
        self.icon.set_tooltip_text("Sonos Stream")
        self.icon.set_visible(True)

        self.menu = self.build_menu()
        self.icon.connect("popup-menu", self.on_popup)
        self.icon.connect("activate", self.on_activate)

    def build_menu(self):
        menu = Gtk.Menu()

        groups = self.data.get("groups", [])
        active_count = sum(1 for g in groups if g.get("active"))

        # Status
        status = Gtk.MenuItem(label=f"Streaming to {active_count} system(s)")
        status.set_sensitive(False)
        menu.append(status)
        menu.append(Gtk.SeparatorMenuItem())

        # Level 1: Groups (systems)
        for group in groups:
            group_name = group.get("name", "Unknown")
            active = group.get("active", False)

            if active:
                # Use a check menu item to show selection
                group_item = Gtk.CheckMenuItem(label=group_name)
                group_item.set_active(True)
                group_item.set_sensitive(True)
                # Prevent toggling
                group_item.connect("toggled", lambda w: w.set_active(True))
            else:
                group_item = Gtk.MenuItem(label=group_name)

            # Level 2: Devices in this group
            devices_menu = Gtk.Menu()
            for dev in group.get("devices", []):
                dev_label = dev.get("name", "Unknown")
                model = dev.get("model", "")
                if model:
                    dev_label += f" ({model})"
                if dev.get("is_coordinator"):
                    dev_label += " ★"

                dev_item = Gtk.MenuItem(label=dev_label)

                # Level 3: Device details
                details_menu = Gtk.Menu()
                for key, val in [
                    ("IP", dev.get("ip")),
                    ("Model", f"{dev.get('model')} ({dev.get('model_number')})"),
                    ("UUID", dev.get("uuid")),
                    ("Serial", dev.get("serial")),
                    ("Software", dev.get("software")),
                    ("Hardware", dev.get("hardware")),
                    ("Coordinator", str(dev.get("is_coordinator", False))),
                    ("Invisible", str(dev.get("invisible", False))),
                    ("Can play", str(dev.get("can_play", False))),
                ]:
                    info = Gtk.MenuItem(label=f"{key}: {val}")
                    info.set_sensitive(False)
                    details_menu.append(info)

                dev_item.set_submenu(details_menu)
                devices_menu.append(dev_item)

            group_item.set_submenu(devices_menu)
            menu.append(group_item)

        menu.append(Gtk.SeparatorMenuItem())

        # Quit
        quit_item = Gtk.MenuItem(label="Quit")
        quit_item.connect("activate", self.on_quit)
        menu.append(quit_item)

        menu.show_all()
        return menu

    def on_popup(self, icon, button, time):
        self.menu.popup(None, None, Gtk.StatusIcon.position_menu, icon, button, time)

    def on_activate(self, icon):
        self.menu.popup(None, None, Gtk.StatusIcon.position_menu, icon, 1,
                        Gtk.get_current_event_time())

    def on_quit(self, _):
        print("QUIT", flush=True)
        Gtk.main_quit()


def main():
    line = sys.stdin.readline().strip()
    if not line:
        sys.exit(1)
    data = json.loads(line)

    signal.signal(signal.SIGTERM, lambda *_: Gtk.main_quit())
    signal.signal(signal.SIGINT, lambda *_: Gtk.main_quit())

    # Exit when parent dies (stdin closes)
    channel = GLib.IOChannel.unix_new(sys.stdin.fileno())
    GLib.io_add_watch(channel, GLib.PRIORITY_DEFAULT, GLib.IOCondition.HUP,
                      lambda *_: (Gtk.main_quit(), False)[-1])

    SonosTray(data)
    Gtk.main()


if __name__ == "__main__":
    main()

File: tray.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"sonos-linux/sonos"
)

type trayApp struct {
	cmd  *exec.Cmd
	quit chan struct{}
}

type trayGroup struct {
	ID      string       `json:"id"`
	Name    string       `json:"name"`
	Active  bool         `json:"active"`
	Devices []trayDevice `json:"devices"`
}

type trayDevice struct {
	Name            string `json:"name"`
	IP              string `json:"ip"`
	Model           string `json:"model"`
	ModelNumber     string `json:"model_number"`
	UUID            string `json:"uuid"`
	Serial          string `json:"serial"`
	Software        string `json:"software"`
	Hardware        string `json:"hardware"`
	IsCoordinator   bool   `json:"is_coordinator"`
	Invisible       bool   `json:"invisible"`
	CanPlay         bool   `json:"can_play"`
}

func startTray(devices []*sonos.Device, streams []deviceStream) *trayApp {
	t := &trayApp{quit: make(chan struct{})}

	trayScript := findTrayScript()
	if trayScript == "" {
		log.Print("tray.py not found, running without system tray")
		return t
	}

	// Build set of active device UUIDs (ones we're streaming to)
	activeUUIDs := make(map[string]bool)
	for _, s := range streams {
		activeUUIDs[s.device.UUID] = true
	}

	// Group devices by GroupID
	groupMap := make(map[string]*trayGroup)
	var groupOrder []string
	for _, dev := range devices {
		g, exists := groupMap[dev.GroupID]
		if !exists {
			g = &trayGroup{ID: dev.GroupID}
			groupMap[dev.GroupID] = g
			groupOrder = append(groupOrder, dev.GroupID)
		}
		td := trayDevice{
			Name:          dev.Name,
			IP:            dev.IP,
			Model:         dev.ModelName,
			ModelNumber:   dev.ModelNumber,
			UUID:          dev.UUID,
			Serial:        dev.SerialNumber,
			Software:      dev.SoftwareVersion,
			Hardware:      dev.HardwareVersion,
			IsCoordinator: dev.IsCoord,
			Invisible:     dev.Invisible,
			CanPlay:       dev.CanPlay(),
		}
		g.Devices = append(g.Devices, td)
		// Name the group after the coordinator
		if dev.IsCoord {
			g.Name = dev.Name
		}
		// Group is active if any of its devices is being streamed to
		if activeUUIDs[dev.UUID] {
			g.Active = true
		}
	}

	var groups []trayGroup
	for _, id := range groupOrder {
		groups = append(groups, *groupMap[id])
	}

	data, _ := json.Marshal(map[string]any{"groups": groups})

	cmd := exec.Command("python3", trayScript)
	cmd.Stderr = os.Stderr

	stdin, err := cmd.StdinPipe()
	if err != nil {
		log.Printf("Failed to create tray stdin pipe: %v", err)
		return t
	}

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		log.Printf("Failed to create tray stdout pipe: %v", err)
		return t
	}

	if err := cmd.Start(); err != nil {
		log.Printf("Failed to start tray: %v", err)
		return t
	}
	t.cmd = cmd

	fmt.Fprintf(stdin, "%s\n", data)

	go func() {
		scanner := bufio.NewScanner(stdout)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "QUIT" {
				close(t.quit)
				return
			}
		}
		select {
		case <-t.quit:
		default:
			close(t.quit)
		}
	}()

	return t
}

func (t *trayApp) stop() {
	if t.cmd != nil && t.cmd.Process != nil {
		t.cmd.Process.Signal(os.Interrupt)
		t.cmd.Wait()
	}
}

func findTrayScript() string {
	if exe, err := os.Executable(); err == nil {
		p := filepath.Join(filepath.Dir(exe), "tray.py")
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	if _, err := os.Stat("tray.py"); err == nil {
		return "tray.py"
	}
	return ""
}
